src/bp/ui-admin/src/Pages/Workspace/Bots/EditStageModal.tsx

```diff
@@ -0,0 +1,205 @@
+
+      setIsLastPipeline(pipeline[pipeline.length - 1].id === id)
+      setLabel(label || '')
+      setAction(isLastPipeline ? 'noop' : action || 'promote_copy')
+      setReviewers(reviewers || [])
+      setMinimumApprovals(minimumApprovals || 0)
+      setReviewSequence(reviewSequence || 'serial')
+    }
+  }, [props.stage])
+
+  useEffect(() => {
+    if (props.workspace) {
+      const { adminRole, pipeline } = props.workspace
+
+      setPipeline(pipeline)
+      props.fetchUsers(adminRole)
+    }
+  }, [])
+
+  const submit = async () => {
+    const {
+      stage: { id }
+    } = props
+    let newPipeline
+
+    if (!pipeline.find(p => p.id === id)) {
+      toastFailure('Could not find the pipeline to save')
+    } else {
+      newPipeline = pipeline.map(p =>
+        p.id !== id
+          ? p
+          : {
+              ...p,
+              label,
+              action,
+              reviewers,
+              minimumApprovals,
+              reviewSequence
+            }
+      )
+    }
+
+    try {
+      await api
+        .getSecured()
+        .post(`/admin/workspaces/${getActiveWorkspace()}/pipeline`, { updateCustom: true, pipeline: newPipeline })
+
+      props.onEditSuccess()
+      closeModal()
+    } catch (error) {
+      toastFailure(`Error while updating pipeline: ${error.message}`)
+    } finally {
+      setProcessing(false)
+    }
+  }
+
+  const closeModal = () => {
+    setProcessing(false)
+    props.toggle()
+  }
+
+  const formatUser = user => {
+    const {
+      email,
+      attributes: { firstname, lastname }
+    } = user
+
+    return {
+      label: firstname || lastname ? `${firstname} ${lastname} · ${email}` : email,
+      value: email
+    }
+  }
+
+  const onReviewersChange = value => {
+    setReviewers(value)
+
+    if (value.length < minimumApprovals) {
+      setMinimumApprovals(value.length)
+    }
+  }
+
+  const { stage, toggle, users } = props
+
+  return (
+    <Dialog isOpen={props.isOpen} icon="undo" onClose={closeModal} transitionDuration={0} title={label}>
+      <div className={Classes.DIALOG_BODY}>
+        <FormGroup label="Label">
+          <Input
+            id="input-label"
+            type="text"
+            value={label}
+            onChange={({ target: { value } }) => setLabel(value)}
+            autoFocus={true}
+          />
+        </FormGroup>
+        {!isLastPipeline && (
+          <FormGroup label="Action">
+            <RadioGroup onChange={({ currentTarget: { value } }) => setAction(value)} selectedValue={action} inline>
+              <Radio label="Copy" value="promote_copy" />
+              <Radio label="Move" value="promote_move" />
+            </RadioGroup>
+          </FormGroup>
+        )}
+        <FormGroup label="Reviewers">
+          <Select
+            id="select-reviewers"
+            isMulti={true}
+            value={reviewers}
+            options={users && users.map(formatUser)}
```

Comment:
If there's a lot of users, it's going to be processed on each render.Maybe process it once and set in the state ?